elytron/WFLY-14735-Management-FORM-authentication.adoc

```diff
@@ -0,0 +1,245 @@
+
+Updates will be required to the https://github.com/hal/console[Admin Console] to support
+possibly being required to redirect to the login form and to make use of the logout endpoint.
+
+Security components developed for this RFE will be contained within the
+https://github.com/wildfly-security/wildfly-elytron[WildFly Elytron] project.
+
+=== Other Interested Projects
+
+None
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+==== General Flow
+
+On the HTTP management interface a new context will be made available at the path `/authentication`,
+when a user connects to the root `/` context of the application server they will be redirected to the
+`/authentication` context.  The `/authentication` context will display a login form for the user 
+to authenticate themselves.
+
+The look and feel of the login FORM will make use of `PatternFly` for consistent with the look and
+feel of the web console.
+
+Under the `/authentication` context there will be two sub contexts `/login` and `/logout`, after 
+a user has input their username and password these will be passed to the first of the sub contexts to
+perform the authentication.  Where authentication is successful a cryptographically signed JWT token 
+will be generated and passed back to the client as a cookie, the login form will then redirect the 
+user to the admin console.
+
+The admin console context is unauthenticated so will load as normal, the `/management` enpoint will 
+use a different HTTP authentication mchanism configuration which will use the `BEARER` HTTP 
+authentication mechanism which will be configured to use the JWT token from the incoming request cookie.
+
+When the user has finished managing the application server they can click the usual Logout link in 
+the admin console which will redirect them to `/authentication/logout`, this context will clear 
+the cookie containing the JWT token.
+
+==== Other Flows
+
+If after the username and password is sent to the `/authentication/login` context the authentication 
+fails the user will remain on the login page and a suitably vague error will be delayed to indicate
+that authentication has failed.
+
+If a user has previously been authenticated and their cookie has expired or the user is using the 
+admin console to connect to a different server it is possible the admin console may sent an invocation
+to the `/management` context without an appropriate cookie, in this case the admin console will need
+to be "notified to" and handle a redirect to the `/authentication` context.
+
+==== elytron-authentication
+
+A new subsystem `elytron-authentication` will be developed, the primary role of this subsystem
```

Comment:
This name seems quite generic for a subsystem that's targeted at a pretty specific use.

Reply:
Even if expanded to deployment use the name is kind of generic, and the elytron subsystem itself deals with authentication too.

All this said I know names are hard and I don't have any suggestions, so please just regard this as a generic 'even though you thought about the name 10 times, please think twice more' request.

Reply:
Be careful I will get the thesaurus out - maybe elytron-form-authentication the problem is if we do get round to fully supporting Fido2 WebAuthn the user no longer interacts with the FORM.  Possibly elytron-http-authentication?

Reply:
I wrote this comment before properly digesting the future enhancements section in the 'Non Requirements'. Those don't eliminate the issue but do give a sense of this subsystem over time more clearly justifying the general purpose name.

I don't think elytron-http-authenication is better as when I pretend not to understand things I think "The elytron subsystem handles http authentication; I know because I see http a lot in its config!" So that's just a more verbose variant of the same problem.

Reply:
One point to raise regarding subsystem naming is the overall naming scheme for elytron subsystems.  For better or for worse our main subsystem is called "elytron" with no further qualifiers, this subsystem is intended to be a verbose subsystem enabling an administrator to build their own policy from the building blocks we provide and through the use of capabilities potentially provided by others.

There is a new family of subsystems considered that have specialist behaviour such as "elytron-tls" and "elytron-authentication" - these are intended to cover a large percentage of possible requirements whilst minimising the complexity of the configuration.  So there is going to duplication, the elytron-* subsystems simplify specific use cases whilst the elytron subsystem provides the complete verbose set of options.

Continuing the naming for this new subsystem, another possibility could be something like "elytron-interactive", or "elytron-dynamic" to indicate it's purpose is for an interactive authentication process.  possibly "elytron-dynamic-form" to indicate the primary purpose is dynamic form authentication?

Reply:
> There is a new family of subsystems considered that have specialist behaviour such as "elytron-tls" and "elytron-authentication" - these are intended to cover a large percentage of possible requirements whilst minimising the complexity of the configuration. So there is going to duplication, the elytron-* subsystems simplify specific use cases whilst the elytron subsystem provides the complete verbose set of options.

From the analysis for `elytron-tls` the subsystem would not seem to be a simplification as the complexity seems the same, just limited to TLS settings.

What will be the content of `elytron-authentication`? Just a set of something like `http-context` resource referencing `security-domain` and handling a list of dynamic attributes for extra FORM fields? Also there is no requirement to have such resources in the basic `elytron` subsystem.

Resources of `elytron-oidc-client` are also not included in the basic `elytron` subsystem.

Regarding the name, do we expect we will add another (HTTP based) authentication methods (than FORM) to the subsystem in future?

Reply:
Then something like `elytron-dynamic-form`, `elytron-form`,  `elytron-authentication-form` could be more suitable than the general `elytron-authentication`.

I suppose in (now unexpected) case of adding another authn method in the future, we would place it into another new subsystem. Because of layers an so on.

Reply:
I like elytron-dynamic-form

It does have a minimal dynamic aspect from the outset, this will then grow in the future.

Reply:
Feature team has agreed the name will be elytron-dynamic-form